File: src/main/java/com/brandonitaly/bedrockskins/client/SkinManager.java
package com.brandonitaly.bedrockskins.client;

import java.util.HashMap;
import java.util.Map;
import net.minecraft.client.Minecraft;

public final class SkinManager {
    private SkinManager() {}

    private static final Map<String, String> playerSkins = new HashMap<>();

    public static void load() {
        playerSkins.clear();
        try {
            var state = StateManager.readState();
            var selected = state.getSelected();
            var client = Minecraft.getInstance();
            var player = client.player;
            if (selected != null && !selected.isEmpty() && player != null) {
                playerSkins.put(player.getUUID().toString(), selected);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static String getLocalSelectedKey() {
        var client = Minecraft.getInstance();
        var localUuid = client.player != null ? client.player.getUUID().toString() : null;
        if (localUuid == null) return null;
        return playerSkins.get(localUuid);
    }

    public static void setSkin(String uuid, String packName, String skinName) {
        String key = packName + ":" + skinName;
        playerSkins.put(uuid, key);
        var client = Minecraft.getInstance();
        var localUuid = client.player != null ? client.player.getUUID().toString() : null;
        if (localUuid != null && localUuid.equals(uuid)) {
            try {
                var favorites = FavoritesManager.getFavoriteKeys();
                StateManager.saveState(favorites, key);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public static void setPreviewSkin(String uuid, String packName, String skinName) {
        String key = packName + ":" + skinName;
        playerSkins.put(uuid, key);
    }

    public static void resetPreviewSkin(String uuid) {
        playerSkins.remove(uuid);
    }

    public static String getSkin(String uuid) {
        return playerSkins.get(uuid);
    }

    public static void resetSkin(String uuid) {
        if (playerSkins.remove(uuid) != null) {
            var client = Minecraft.getInstance();
            var localUuid = client.player != null ? client.player.getUUID().toString() : null;
            if (localUuid != null && localUuid.equals(uuid)) {
                try {
                    var favorites = FavoritesManager.getFavoriteKeys();
                    StateManager.saveState(favorites, null);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: src/main/kotlin/com/brandonitaly/bedrockskins/client/SkinManager.kt
package com.brandonitaly.bedrockskins.client

import net.minecraft.client.MinecraftClient

object SkinManager {
    // Map of Player UUID (String) to Skin Key (packName:skinName) for runtime only
    private val playerSkins = mutableMapOf<String, String>()
    // Load the saved local player's skin from the shared state file and store it in the runtime map
    fun load() {
        playerSkins.clear()
        try {
            val state = StateManager.readState()
            val selected = state.selected
            val client = MinecraftClient.getInstance()
            val player = client.player
            if (!selected.isNullOrEmpty() && player != null) {
                playerSkins[player.uuid.toString()] = selected
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    // Return the selected skin key for the local player, if any.
    fun getLocalSelectedKey(): String? {
        val client = MinecraftClient.getInstance()
        val localUuid = client.player?.uuid?.toString() ?: return null
        return playerSkins[localUuid]
    }

    fun setSkin(uuid: String, packName: String, skinName: String) {
        val key = "$packName:$skinName"
        playerSkins[uuid] = key
        val client = MinecraftClient.getInstance()
        val localUuid = client.player?.uuid?.toString()
        if (localUuid != null && localUuid == uuid) {
            // Persist combined state (favorites + selected)
            try {
                val favorites = FavoritesManager.getFavoriteKeys()
                StateManager.saveState(favorites, key)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    // Set a preview-only skin mapping without persisting state.
    fun setPreviewSkin(uuid: String, packName: String, skinName: String) {
        val key = "$packName:$skinName"
        playerSkins[uuid] = key
    }

    // Remove a preview-only skin mapping without persisting state.
    fun resetPreviewSkin(uuid: String) {
        playerSkins.remove(uuid)
    }

    fun getSkin(uuid: String): String? {
        return playerSkins[uuid]
    }

    fun resetSkin(uuid: String) {
        if (playerSkins.remove(uuid) != null) {
            val client = MinecraftClient.getInstance()
            val localUuid = client.player?.uuid?.toString()
            if (localUuid != null && localUuid == uuid) {
                try {
                    val favorites = FavoritesManager.getFavoriteKeys()
                    StateManager.saveState(favorites, null)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }
}
